src/UCTNode.h

```diff
@@ -24,20 +24,97 @@
 #include <atomic>
 #include <memory>
 #include <vector>
+#include <cassert>
+#include <cstring>
 
 #include "GameState.h"
 #include "Network.h"
 #include "SMP.h"
 
+class UCTNode;
+
+// 'lazy-initializable' version of std::unique_ptr<UCTNode>.
+// When a UCTNodePointer is constructed, the constructor arguments are stored instead
+// of constructing the actual UCTNode instance.  Later when the UCTNode is needed,
+// the external code calls expand() which actually constructs the UCTNode.
+// Basically, this is a 'tagged union' of:
+//  - std::unique_ptr<UCTNode> pointer;
+//  - std::pair<float, std::int16_t> args;
+
+// WARNING : expand() is not thread-safe and hence has to be protected by an external lock.
+
+class UCTNodePointer {
+private:
+    // the raw storage used here.
+    // if bit 0 is 0, m_data is the actual pointer.
+    // if bit 0 is 1, bit [31:16] is the vertex value, bit [63:32] is the score.
+    // (I really wanted to use C-style bit fields and unions, but those aren't portable)
+    mutable uint64_t m_data = 1;
+
+    bool _expanded() const {
+        return (m_data & 1ULL) == 0;
+    } 
+    UCTNode * _readptr() const {
+        assert(_expanded());
+        return reinterpret_cast<UCTNode*>(m_data);
+    }
+
+    std::int16_t _vertex() const {
+        assert(!_expanded());
+        std::uint16_t x = static_cast<std::uint16_t>( (m_data >> 16) & 0xffff);
+        return static_cast<std::int16_t>(x);
+    }
+    float _score() const {
+        static_assert(sizeof(float) == sizeof(std::uint32_t), "This code assumes floats are 32-bit");
```

Comment:
Why not compare to 4 directly?

Reply:
Wanted to be explicit that I was intended to copy this to a std::uint32_t but well, doesn't really matter.  I'll put a 4 instead.